fuel-core/src/executor.rs

```diff
@@ -2449,6 +2574,281 @@ mod tests {
+        .copied()
+        .collect();
+
+        let modify_balance_and_state_tx = TxBuilder::new(2322)
+            .gas_limit(10000)
+            .coin_input(AssetId::zeroed(), 10000)
+            .start_script(script, script_data)
+            .contract_input(contract_id)
+            .coin_input(asset_id, transfer_amount)
+            .contract_output(&contract_id)
+            .build()
+            .transaction()
+            .clone();
+        let db = &mut Database::default();
+
+        let executor = Executor {
+            database: db.clone(),
+            config: Config {
+                utxo_validation: false,
+                ..Config::local_node()
+            },
+        };
+
+        let block = PartialFuelBlock {
+            header: PartialFuelBlockHeader {
+                consensus: FuelConsensusHeader {
+                    height: 1u64.into(),
+                    ..Default::default()
+                },
+                ..Default::default()
+            },
+            transactions: vec![create.into(), modify_balance_and_state_tx.into()],
+        };
+
+        let ExecutionResult {
+            block, tx_status, ..
+        } = executor.execute(ExecutionBlock::Production(block)).unwrap();
+
+        let empty_state = Bytes32::from(*empty_sum());
+        let executed_tx = block.transactions()[2].as_script().unwrap();
+        assert!(matches!(
+            tx_status[2].status,
+            TransactionStatus::Success { .. }
+        ));
+        assert_eq!(executed_tx.inputs()[0].state_root(), Some(&empty_state));
+        assert_eq!(executed_tx.inputs()[0].balance_root(), Some(&empty_state));
+        // Roots should be different
+        assert_ne!(
+            executed_tx.inputs()[0].state_root(),
+            executed_tx.outputs()[0].state_root()
+        );
+        assert_ne!(
+            executed_tx.inputs()[0].balance_root(),
+            executed_tx.outputs()[0].balance_root()
+        );
+    }
+
+    #[test]
+    fn foreign_transfer_should_not_affect_balance_root() {
+        // The foreign transfer of tokens should not affect the balance root of the transaction.
+        let mut rng = StdRng::seed_from_u64(2322u64);
+
+        let (create, contract_id) = create_contract(vec![], &mut rng);
+
+        let transfer_amount = 100 as Word;
+        let asset_id = AssetId::from([2; 32]);
+        let mut foreign_transfer = TxBuilder::new(2322)
+            .gas_limit(10000)
+            .coin_input(AssetId::zeroed(), 10000)
+            .start_script(vec![Opcode::RET(1)], vec![])
+            .coin_input(asset_id, transfer_amount)
+            .coin_output(asset_id, transfer_amount)
+            .build()
+            .transaction()
+            .clone();
+        if let Some(Output::Coin { to, .. }) = foreign_transfer
+            .as_script_mut()
+            .unwrap()
+            .outputs_mut()
+            .last_mut()
+        {
+            *to = Address::try_from(contract_id.as_ref()).unwrap();
```

Comment:
@Voxelot I added this test but not sure what our expectation is. Maybe we want to reject transactions that try sending tokens to the contract via outputs(not the script body).

Reply:
Hmm yeah, we could do this but it would add extra overhead to every transfer to check if the output address is a contract id.